File: 18-4Sum.py
nums = [-3,-2,-1,0,0,1,2,3]
target = 0

def fourSum(nums, target):
    nums.sort()
    N = len(nums)
    nums_4sum = []
    nums_set = set(nums)
    nums_dic = {}
    res_set = set([])
    res = []
    nums_sets = []
    res_dic = {}

    if  N < 4:
        return []

    for i in range(0, N):
        nums_sets.append(set(nums[i:]))

    st = N - 1
    ed = 0
    for i in range(0, N - 3):
        min = sum(nums[i: i + 4])
        max = nums[i] + nums[N-1] + nums[N-2] + nums[N-3]
        if target >= min and target <= max:
            if st > i:
                st = i
            if ed < i:
                ed = i
    if st > ed:
        return []

    '''
    for i in range(st, ed + 1):
        a = nums[i]
        for j in range(i + 1, N - 2):
            b = nums[j]
            for k in range(j + 1, N - 1):
                c = nums[k]
                d = target - a - b - c
                #print(a, b, c, d)
                if d in nums_sets[k + 1]:
                    res_set.add(str([a, b, c, d]))
                    #res_dic[str([a, b, c, d])] = [a, b, c, d]
    '''
    
    for i in range(st, ed + 1):
        a = nums[i]
        for j in range(i + 1, N - 2):
            b = nums[j]
            cd = target - a - b
            low = j + 1
            high = N - 1
            while low < high:
                c = nums[low]
                d = nums[high]
                if c + d == cd:
                    res_set.add(str([a, b, c, d]))
                    low += 1
                    high -= 1
                elif c + d < cd:
                    low += 1
                else:
                    high -= 1
            if low < high:
                res_set.add(str([a, b, c, d]))
    
    for i in res_set:
        res.append(eval(i))

    return res

print(fourSum(nums, target))  

File: 18-4Sum.c
#include <stdio.h>
#include <stdlib.h>

int comp(const void *a, const void *b){
	return *(int *)a - *(int *)b;
}

void add(int *arr, int **res, int *returnSize){
	int l;
	int dup = 0;
    /* O(n) */
	for(l = 0; l < *returnSize; l++){
		if(res[l][0] == arr[0] && res[l][1] == arr[1] &&
		res[l][2] == arr[2] && res[l][3] == arr[3]){
			dup = 1;
			break;
		}
	}

	if(!dup){
		res[*returnSize] = arr;
		(*returnSize)++;
		//printf("%d %d %d %d\n", arr[0], arr[1], arr[2], arr[3]);
	}else{
		free(arr);
	}
}

int** fourSum(int* nums, int numsSize, int target, int* returnSize) {
	int **res;
	int i, j, k, l;
	int a, b, c, d, cd;
	int dup;
	int *arr;
	int st, ed;
	int max, min;
	int low, high;
	
	*returnSize = 0;
    res = (int **)malloc(sizeof(int *)*1000);
	qsort(nums, numsSize, sizeof(int), comp);
	
	if(numsSize < 4){
		*returnSize = 0;
		return res;
	}
	
    /* Optimize */
	st = numsSize - 1;
	ed = 0;
	for(i = 0; i < numsSize - 3; i++){
		min = nums[i] + nums[i + 1] + nums[i + 2] + nums[i + 3];
		max = nums[i] + nums[numsSize - 1] + nums[numsSize - 2] + nums[numsSize - 3];
		if(target >= min && target <= max){
			if(st > i) st = i;
			if(ed < i) ed = i;
		}
	}

	if(st > ed){
		*returnSize = 0;
		return res;
	}
    
    /* O(n^3) */
	for(i = st; i < ed + 1; i++){
		a = nums[i];
		for(j = i + 1; j < numsSize - 2; j++){
			b = nums[j];
			cd = target - a - b;
			low = j + 1;
			high = numsSize - 1;
			while(low < high){
				c = nums[low];
				d = nums[high];
				if(c + d == cd){
					arr = (int *)malloc(sizeof(int) * 4);
					arr[0] = a;
					arr[1] = b;
					arr[2] = c;
					arr[3] = d;
					add(arr, res, returnSize);
					low += 1;
					high -= 1;
				}else if(c + d < cd){
					low += 1;
				}else
					high -= 1;
			}
		}
	}
	return res;
}

int main(){
	int nums[] = {1, 0, -1, 0, -2, 2};
	int returnSize = 0;
	int numsSize = 6;
	int target = 0;
	
	fourSum(nums, numsSize, target, &returnSize);
	return 0;
}
